implementation of chinese remainder theorem over cartesian product of vectors

mijo2 opened the pull request:
#### References to other Issues or PRs

Fixes #18428  and incorporates PR #10949

#### Brief description of what is fixed or changed

--> Added the implementation of chinese remainder theorem over cartesian product(crt_cartesian) of vectors which was added in PR #10949
--> Added test cases to check if the function crt_cartesian function implemented works properly in test_modular.py

#### Other comments

Example of using crt_cartesian:
```
>> from sympy.ntheory.modular import crt_cartesian
>> crt_cartesian([[3,5], [3,7]], [7, 11])
[3, 73, 47, 40]
>> crt_cartesian([[1, 5], [4, 7], [6, 8]], [6, 11, 13])
[565, 697, 799, 73, 851, 125, 227, 359]
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* ntheory
  * Added new function that implements chinese remainder theorem over cartesian product of vectors 
<!-- END RELEASE NOTES -->

mijo2 committed:
ntheory: Addition of chinese remainder theorem over cartesian product of vectors

In this commit, the changes introuduced in PR for issue #10949 has been added.
The issue comprised of implementing chinese remainder theorem over cartesian product
of vectors. So, the changes that were introuduced in that commit has been added in
this commit which includes implementation and test cases of the theorem.

Example:

>>> from sympy.ntheory.modular import crt_cartesian
>>> crt_cartesian([[3,5], [3,7]], [7, 11])
[3, 73, 47, 40]
>>> [(p%7, p%11) for p in [3, 40, 47, 73]]
[(3, 3), (5, 7), (5, 3), (3, 7)]

namannimmo10 commented:
Please address [this](https://github.com/sympy/sympy/pull/18429#discussion_r369420114) comment also.

smichr commented:
So what is an example of a problem where one would want to do this? That might suggest the most natural argument list for the function.

smichr commented:
The release notes need only tell that a new function has been added, not that you tested it. Think from the perspective of someone that is wondering what has changed in the next release of SymPy: they assume you have tested what was added. Also, don't reference a PR in the release notes since the source of the work is irrelevant to the user.

mijo2 commented:
Ok @smichr, I have added the test case you mentioned [here](https://github.com/sympy/sympy/pull/18429#discussion_r369429289). So now, I have to add a check routine in the function and change the arguments to a dictionary type whose keys are the moduli and whose values are the corresponding remainders correct?

mijo2 commented:
@smichr when i tried running ```crt_cartesian``` function with moduli ```[7, 14]``` I got an error which was mainly because of the reason that there is a ```mod_inverse(m, mod[i])``` line in the implementation which throws ```ValueError: inverse of 7 (mod 14) does not exist```. While when I run ```crt``` function with moduli ```[7, 14]``` I get ```None``` as the result since the way ```gf_crt``` function computes is different from the method used in ```crt_cartesian```. What should I do now? Try to fix the ```crt_cartesian``` method so that it also returns ```None``` when any pair of moduli are co-prime?

mijo2 commented:
So @smichr , for the check I came up with something like below:
```
        try:
            s = mod_inverse(m, mod[i])
        except ValueError:
            return None
```
I am as of now assuming that if any pair of moduli are co-prime, then None has to be returned by ```crt_cartesian``` function. But I don't know as of now how to change this implementation to fully incorporate the check functionality that is used in ```crt``` function which allows us to get the results even when moduli aren't co-prime.

smichr commented:
> The above block can be before the main computation in `crt_cartesian` function

Since computation of the inverse is required here but not in `crt` then it makes sense to just return None in a try/except block around that calculation. No need for a flag -- you won't be able to do the calculation if the condition is not met. dosctring can be edited to say that None is returned if moduli are not pairwise-prime.

mijo2 commented:
Ok @smichr so I will do something like below - 
```
...
rem2 = []
        try:
            s = mod_inverse(m, mod[i])
        except ValueError:
            return None
        _m = m
...

mijo2 commented:
Also, I have added this test case 
```assert crt_cartesian([[4, 2], [5, 7], [1, 3]], [6, 11, 14]) == None``` for checking if the function returns None or not

mijo2 committed:
ntheory: Adding a functionality in crt_cartesian function

In crt_cartesian function, following changes have been made:
1. The function now returns None if any pair of moduli aren't relatively prime
2. The docstring of the function was changed to allow other users to know about the
   functionality added above
3. A test case was added in test_modular.py that checks if the functionality added
   in this commit works properly or not.

Example:

>> from sympy.ntheory.modular import crt_cartesian
>>> crt_cartesian([[4, 2], [5, 7], [1, 3]], [6, 11, 14])
>>> crt_cartesian([[1, 5], [4, 7], [6, 8]], [6, 11, 13])
[565, 697, 799, 73, 851, 125, 227, 359]

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,39 @@ def combine(c1, c2):
             return symmetric_residue(n, m), m
         return n, m
+
+
```
The doctest is not being tested, currently. Some friendly suggestions (which don't include changing the arg list) are:
```diff
diff --git a/sympy/ntheory/modular.py b/sympy/ntheory/modular.py
index e5b14a9..92ce21a 100644
--- a/sympy/ntheory/modular.py
+++ b/sympy/ntheory/modular.py
@@ -257,35 +257,36 @@ def combine(c1, c2):
 
 def crt_cartesian(rem, mod):
     """
-    Crt over a cartesian product of vectors.given a vector of moduli,
-    and for each modulus given a vector of remainders, instead of usual one remainder in crt.
-    It outputs solutions that leave a remainder matching one of the remainders
-    of the vector of remainders corresponding to a modulus.
-    Assuming that moduli are pairwise relatively prime.
+    Return a list of values whose elements give the remainders
+    in the Cartesian product of the supplied remainders when divided
+    by each of the supplied moduli. None is returned if the moduli
+    are not pairwise-prime.
+
     Examples
     ========
+
     >>> from sympy.ntheory.modular import crt_cartesian
-    >>> crt_cartesian([[3,5], [3,7]], [7, 11])
+    >>> from sympy.utilities.iterables import cartes
+    >>> rem = [[3, 5], [3, 7]]
+    >>> crt_cartesian(rem, [7, 11])
     [3, 73, 47, 40]
-    >>> [(p%7, p%11) for p in [3, 40, 47, 73]]
-    [(3, 3), (5, 7), (5, 3), (3, 7)]
+    >>> [(p%7, p%11) for p in _]
+    [(3, 3), (3, 7), (5, 3), (5, 7)]
+    >>> assert _ == list(cartes(*rem))
+
     """
-    if len(mod) > len(rem):
-        raise ValueError("Too few remainders")
     if len(mod) == 0:
-        raise ValueError("Moduli Vector can't be empty")
+        raise ValueError("There must be at least 1 modulus.")
+    if len(mod) != len(rem) or not all(rem):
+        raise ValueError(
+            "There should be a list of remainders for each modulus.")
     m = mod[0]
     R = rem[0]
-    for i in range(1,len(mod)):
-        rem2 = []
-        s = mod_inverse(m, mod[i])
-        _m = m
+    for i in range(1, len(mod)):
+        try:
+            mi = m*mod_inverse(m, mod[i])
+        except ValueError:
+            return None  # mod_inverse fails if values are not coprime
         m *= mod[i]
-        for elem in R:
-            for k in rem[i]:
-                r = elem
-                r += _m*s*(k - r)
-                r %= m
-                rem2.append(r)
-        R = rem2
+        R = [(r + mi*(ri - r)) % m for r in R for ri in rem[i]]
     return R
```

mijo2 replied:
```R = [(r + mi*(ri - r)) % m for r in R for ri in rem[i]]``` this gave incorrect results.
I tried to get the correct expression and found something like below:
 ```
        prev_m = m
        m *= mod[i]
        R = [(r + prev_m*mi*(ri - r)) % m for r in R for ri in rem[i]]
```

smichr replied:
Yes, I left off the `m` from the `mi` caluclation.

mijo2 committed:
ntheory: Improving the quality of the code in crt_cartesian function

In this commit, the docstring and the code for the function was tweaked
to make it more presentable and efficient.

namannimmo10 commented:
Hopefully, the build for the failing test will start again soon.

mijo2 commented:
@namannimmo10 so if the build doesn't restart, I have to submit another commit after a while right?

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,38 @@ def combine(c1, c2):
+    R = rem[0]
+    for i in range(1,len(mod)):
+        try:
+            mi = mod_inverse(m, mod[i])
```
```suggestion
            mi = mod_inverse(m, mod[i])*m
```

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,38 @@ def combine(c1, c2):
+            "There should be a list of remainders for each modulus.")
+    m = mod[0]
+    R = rem[0]
+    for i in range(1,len(mod)):
```
```suggestion
    for i in range(1, len(mod)):
```

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,38 @@ def combine(c1, c2):
+            mi = mod_inverse(m, mod[i])
+        except ValueError:
+            return None  # mod_inverse fails if values are not coprime
+        prev_m = m
```
```suggestion
```

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,38 @@ def combine(c1, c2):
+            return None  # mod_inverse fails if values are not coprime
+        prev_m = m
+        m *= mod[i]
+        R = [(r + prev_m*mi*(ri - r)) % m for r in R for ri in rem[i]]
```
```suggestion
        R = [(r + mi*(ri - r)) % m for r in R for ri in rem[i]]
```

smichr commented on sympy/ntheory/tests/test_modular.py:
```diff
@@ -32,3 +32,11 @@ def test_modular():
     raises(
         ValueError, lambda: solve_congruence(*list(zip([3, 4, 2], [12.1, 35, 17]))))
+    assert crt_cartesian([[3, 5], [3, 7]], [7, 11]) == [3, 73, 47, 40]
+    assert crt_cartesian([[1, 5], [4, 7], [6, 8]], [6, 11, 13]) == [565, 697, 799, 73, 851, 125, 227, 359]
```
```suggestion
    assert crt_cartesian([[1, 5], [4, 7], [6, 8]], [6, 11, 13]) == [
        565, 697, 799, 73, 851, 125, 227, 359]
```

smichr replied:
(and similar for the other long lines)

smichr commented on sympy/ntheory/tests/test_modular.py:
```diff
@@ -32,3 +32,11 @@ def test_modular():
+    assert crt_cartesian([[1, 5], [4, 7], [6, 8]], [6, 11, 13]) == [565, 697, 799, 73, 851, 125, 227, 359]
+    assert crt_cartesian([[5, 3], [7, 9], [3, 7]], [7, 11, 13]) == [887, 579, 614, 306, 458, 150, 185, 878]
+    assert crt_cartesian([[11, 51], [54, 72], [16, 38]], [67, 79, 43]) == [196053, 111365, 129790, 45102, 189259, 104571, 122996, 38308]
+    assert crt_cartesian([[4, 2], [5, 7], [1, 3]], [6, 11, 14]) == None
```
```suggestion
    assert crt_cartesian([[4, 2], [5, 7], [1, 3]], [6, 11, 14]) is None
```

smichr commented on sympy/ntheory/tests/test_modular.py:
```diff
@@ -32,3 +32,11 @@ def test_modular():
     assert solve_congruence(*list(zip((1, 1, 2), (3, 2, 4)))) is None
     raises(
         ValueError, lambda: solve_congruence(*list(zip([3, 4, 2], [12.1, 35, 17]))))
+    assert crt_cartesian([[3, 5], [3, 7]], [7, 11]) == [3, 73, 47, 40]
```
```suggestion
    assert crt_cartesian([[3, 5], [3, 7]], [7, 11]) == [3, 73, 47, 40]
    assert crt_cartesian([[3], [3, 7]], [7, 11]) == [3, 73]
```

mijo2 committed:
ntheory: Small changes in the files modular.py and test_modular.py

Improvement of code quality according to the standards of the community,
along with addition of new test cases and changing code like
== None to is None as it is used generally by the Python community.
Less number of local variables are used to improve efficiency of the
code.

mijo2 commented:
@smichr I have added your suggestions in the latest commit(4th).

smichr committed:
check results

smichr commented:
Can you give an example of use case? That will help guide the choice of API.

smichr committed:
Update test_modular.py

mijo2 commented:
@smichr I can't think of any particular use case as of now as I have started contributing to this community very recently. Also, I am unaware of the API as well.

mijo2 commented on sympy/ntheory/tests/test_modular.py:
```diff
@@ -1,4 +1,5 @@
-from sympy.ntheory.modular import crt, crt1, crt2, solve_congruence, crt_cartesian
+from sympy.ntheory.modular import crt, crt1, crt2, solve_congruence, \
+    crt_cartesian
```
@smichr I just tested the code, I think you forgot to import ```cartes``` from ```utilities.iterables```. I will add it in the next commit after running the tests

mijo2 committed:
ntheory: Fixing import errors from the last commit

cartes wasn't imported in test_modular.py which was causing
the failure of some test cases. This import error has been
fixed in this commit

smichr committed:
Update modular.py

asmeurer commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,40 @@ def combine(c1, c2):
+    Return a list of values whose elements give the remainders
+    in the Cartesian product of the supplied remainders when divided
+    by each of the supplied moduli. None is returned if the moduli
+    are not pairwise-prime.
```
It seems better to raise an exception in this case.

asmeurer replied:
It would help to spell this out as a formula.

mijo2 replied:
@asmeurer , if we have to raise an exception, then we can do something like below:
```
        try:
            mi = mod_inverse(m, mod[i])*m
        except:
            raise ValueError("two or more moduli aren't co-prime")  # mod_inverse fails if values are not coprime
```

mijo2 replied:
Also, the test cases will have to be updated accordingly

mijo2 replied:
@smichr said that we can return None when any pair of moduli aren't co-prime like done in ```crt``` function. So what should we do? Either update both ```crt``` and ```crt_cartesian``` to raise ValueError or just return None as is in both the functions

asmeurer replied:
I would just let the exception from mod_inverse bubble up. 

I didn't realize that crt already returns None. Returning a different type in the case of error is not very Pythonic. It's better to raise an exception. Otherwise the user has to check the type, and they also don't get any sort of error message.

mijo2 replied:
Ok so should I include a ValueError exception in ```crt``` as well?

mijo2 replied:
@asmeurer So for ValueError exception, we can add a function in ```modular.py``` that takes in two parameters namely ```mods``` which is list of moduli and a moduli ```m``` when a ValueError exception has been raised by the routine.

Now this function which takes ```mods``` and ```m``` as it's arguments will return a pair of moduli which aren't co-prime and in this pair, one of the moduli will be m. This pair can be printed as an output so that the user gets a sense of where the program is going wrong.

This function can be used by both the ```crt``` and ```crt_cartesian``` functions.

mijo2 replied:
@asmeurer When I was testing ```crt``` function, I found out that when any moduli is ```1```, the function still returns an output though it is supposed to be not correct and should raise a ValueError exception IMO.

asmeurer commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +253,40 @@ def combine(c1, c2):
+    >>> from sympy.utilities.iterables import cartes
+    >>> rem = [[3, 5], [3, 7]]
+    >>> crt_cartesian(rem, [7, 11])
+     [3, 73, 47, 40]
```
There is an extra space here.

mijo2 replied:
I will fix this.

mijo2 committed:
ntheory: Added Value Error exceptions

When moduli aren't co-prime, ValuError exception will take place
in both crt and crt_cartesian functions.

Along with that, test cases were modified and added to verify the
functionality added in this commit

mijo2 commented:
@asmeurer @smichr In the latest commit, i have edited the ```crt``` function to return ValueError whenever moduli are co-prime. You can review it as the ```crt``` function is quite an old one.

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -54,17 +54,7 @@ def crt(m, v, symmetric=False, check=True):
        >>> [639985 % m for m in [99, 97, 95]]
        [49, 76, 65]
 
-    If the moduli are not co-prime, you may receive an incorrect result
```
This warning still applies and shouldn't be deleted. Just modify the `crt([12, 6, 17], [3, 4, 2])` example to show the ValueError
```python
>>> crt([12, 6, 17], [3, 4, 2])
Traceback etc...
...
ValueError: etc...
```
i.e. the result will be 3 lines: the first and last lines of the error message with "..." as the 2nd line.

smichr replied:
Move your comment about the error to precede the example.

mijo2 replied:
But will it pass the doctest if I put something like above in the example of docstring? As the doctest carries test of its own

asmeurer replied:
I think @smichr means to replace the `etc...` parts with what is actually printed. The line with `...` will take care of everything between the first and last lines of the error output in the doctest runner.

mijo2 replied:
Ok @asmeurer got it. Thanks!

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -283,10 +274,7 @@ def crt_cartesian(rem, mod):
-            mi = mod_inverse(m, mod[i])*m
-        except ValueError:
-            return None  # mod_inverse fails if values are not coprime
+        mi = mod_inverse(m, mod[i])*m
```
```suggestion
        mi = mod_inverse(m, mod[i])*m  # fails if `mod` not pairwise-prime
```

mijo2 committed:
ntheory: Added small changes to modular.py

A small comment about mod_inverse and an example about
ValueError was added in docstring of crt function

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -54,17 +54,12 @@ def crt(m, v, symmetric=False, check=True):
-    if you use ``check=False``:
+    If the moduli are not co-prime, you will receive a ValueError.
 
-       >>> crt([12, 6, 17], [3, 4, 2], check=False)
```
please restore these 4 lines to show that, with check False, the wrong result might be obtained

smichr committed:
show incorrect use of check=False in docstring

smichr commented:
It looks ready to me. Let's see if there are any other comments.

czgdp1807 commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +255,37 @@ def combine(c1, c2):
+    in the Cartesian product of the supplied remainders when divided
+    by each of the supplied moduli. If the moduli are not co-prime,
+    you will receive a ValueError.
+
```
IMO, the following section should be included in the docstring,
```python
Parameters
=========

parameter_name: type_of_parameter
    Explanation for the parameter, what value it defaults, etc.
```

mijo2 replied:
This section is missing not only in ```crt_cartesian``` but also in ```crt```. I will include them in both

czgdp1807 commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +255,37 @@ def combine(c1, c2):
+    [3, 73, 47, 40]
+    >>> [(p%7, p%11) for p in _]
+    [(3, 3), (3, 7), (5, 3), (5, 7)]
+    >>> assert _ == list(cartes(*rem))
```
I think `assert` shouldn't be used in `Examples`. Instead show them the outputs of the function.

sylee957 replied:
Yes, you can delete this line entirely if this does not look good.
The line above already asserts that the result is correct

mijo2 replied:
Ok @czgdp1807 @sylee957 will delete that line.

czgdp1807 commented:
Other than some minor comments in documentation, LGTM. Will merge if the docs is improved and no one objects for sometime. Thanks.

mijo2 committed:
ntheory: Added parameter details about crt and crt_cartesian

Parameter details about crt and crt_cartesian was added in this
commit in modular.py. Along with that, a example was omitted from
the docstring of crt_cartesian since it was redundant

mijo2 committed:
ntheory: Fixed the underline of title in crt

A small bug of small underline was fixed in this commit.

mijo2 committed:
ntheory: Fixed the underline of title in crt

A small bug of small underline was fixed in this commit.

smichr commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +271,47 @@ def combine(c1, c2):
+    ==========
+
+    rem: 2D list
+        Consists of list of lists whose cartesian product gives
```
"List of lists, each containing the known remainders with respect to each modulus."

smichr committed:
Update test_modular.py

smichr committed:
Update modular.py

smichr committed:
Update modular.py

smichr commented:
My remaining question is whether it makes sense to (as in crt) return `R, m` instead of only `R`. The `m` tells the periodicity of the answer (i.e. for `R = [a, b]` then `[a+m, b+m]` is a congruent answer).

mijo2 commented:
@smichr as done in ```crt``` function we can make it return both ```R``` and ```m```. I can add this in a new commit.

smichr commented:
A use case was in process at #10949 and @abhinav28071999 should probably take a look because it was an alternate(?) method of computing the `nthroot_mod` with a composite modulus.

abhinav28071999 commented:
> A use case was in process at #10949 and @abhinav28071999 should probably take a look because it was an alternate(?) method of computing the `nthroot_mod` with a composite modulus.

I am not sure if that algorithm is correct moreover it is not complete because there is a line `#TODO implement _nthroot_mod_prime for general p**k` and no reference to algorithm is made, i don't know if it is correct or not?

smichr commented:
> I am not sure if that algorithm is correct

sorry...I meant #10946, esp lines [from here](https://github.com/sympy/sympy/pull/10946/files#diff-d43e7fc1bd11e75f0bc7baf255de6605R780)

abhinav28071999 commented:
@smichr Yes it should be correct, but it only works when factors are of form  `2**a` but when factors include like 3**2 or 5**4, there is no implementation for this. Check like `865` .  I can do a timing test to see which one is more efficient.

smichr commented:
> when factors are of form `2**a`

I see what you mean. 

Regarding this PR. I think we need a real-life use-case demonstration before this goes in to get it out of the YAGNI status.

mijo2 commented:
@czgdp1807 I think I just added the implementation introduced in #10949 and tried to fix #18428 but I know little about the use case of this algorithm @smichr. I would try to find more about it and get back to you.

czgdp1807 commented:
@mijo2 Any updates on this?

mijo2 commented:
@czgdp1807 I don't have updates as I wasn't able to find much use cases for this implementation yet.

czgdp1807 commented:
Adding `Could Close` label as there aren't any popular use cases of this algorithm. If someone finds any just comment below.

oscargus commented on sympy/ntheory/modular.py:
```diff
@@ -253,3 +271,47 @@ def combine(c1, c2):
         return n, m
+
+
+def crt_cartesian(rem, mod):
```
Even though the original input order is bit non-intuitive, I think swapping it for `crt_cartesian` is worse.
